### PR TITLE
Use 100 Levels for MySensors Dimmers instead of 16

### PR DESCRIPTION
**Pull request by LemarinelNet**
Here is my try to MySensorsBase.cpp.

I've tried it with my MySensor Dimmer, it works. 100 levels are sent, and On/Off states are still handled.
Haven't tried anything else, but seems to work.

### CODE REVIEW

**Comment by gizmocuz** [comment]
Thank you very much for the patch !
Before i merge, is it possible to try the 'others' if they are still working ?
There is an excellent MySensors Example project, called 'MockMySensor'
In here, you can enable a sensor type, like a door sensor,  and test/debug this

**Comment by LemarinelNet** [comment]
Ok, I'll do some testings this morning

**Comment by LemarinelNet** [comment]
Made a few tests : 
- Lux sensor
- Binary switch (On/Off)
- Dimmer

Everything is OK.

File: hardware/MySensorsBase.cpp [hunk 1]
```diff
@@ -1088,7 +1088,8 @@ void MySensorsBase::UpdateSwitchLastUpdate(const unsigned char Idx, const int Su
 	char szIdx[10];
 	sprintf(szIdx, "%X%02X%02X%02X", 0, 0, 0, Idx);
 	std::vector<std::vector<std::string> > result;
-	result = m_sql.safe_query("SELECT ID FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeLighting2), int(sTypeAC));
+	// LLEMARINEL : #1312  Changed pTypeLighting2 to pTypeGeneralSwitch
+	result = m_sql.safe_query("SELECT ID FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeGeneralSwitch), int(sTypeAC));
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp:1095]: When using pTypeGeneralSwitch, you should use it's subtypes (defined in hardwaretypes.h), for example sSwitchTypeAC
Not sTypeAC

File: hardware/MySensorsBase.cpp [hunk 2]
```diff
@@ -1141,13 +1142,13 @@ void MySensorsBase::UpdateRGBWSwitchLastUpdate(const int NodeID, const int Child
...
-	double rlevel = (15.0 / 100)*Level;
-	int level = int(rlevel);
+	// LLEMARINEL : #1312  Changed to use as pTypeGeneralSwitch : do not constrain to 16 steps anymore but 100 :
+	int level = int(Level);
 
 	char szIdx[10];
 	sprintf(szIdx, "%X%02X%02X%02X", 0, 0, 0, Idx);
 	std::vector<std::vector<std::string> > result;
-	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeLighting2), int(sTypeAC));
+	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeGeneralSwitch), int(sTypeAC));
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp]: You need to use one of the subtypes of pTypeGeneralSwitch, like sSwitchTypeAC

File: hardware/MySensorsBase.cpp [hunk 3]
```diff
@@ -1169,11 +1170,13 @@ void MySensorsBase::UpdateSwitch(const _eSetType vType, const unsigned char Idx,
...
-	//Send as Lighting 2
+	// LLEMARINEL : #1312  Changed to use as pTypeGeneralSwitch
+	// --- Send as Lighting 2
 	tRBUF lcmd;
 	memset(&lcmd, 0, sizeof(RBUF));
 	lcmd.LIGHTING2.packetlength = sizeof(lcmd.LIGHTING2) - 1;
-	lcmd.LIGHTING2.packettype = pTypeLighting2;
+	lcmd.LIGHTING2.packettype = pTypeGeneralSwitch;
+	//lcmd.LIGHTING2.packettype = pTypeLighting2;
 	lcmd.LIGHTING2.subtype = sTypeAC;
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp]: pTypeGeneralSwitch should not be send like this, see for an example RFLinkBase.cpp
CRFLinkBase::SendSwitchInt
You need to use _tGeneralSwitch

**Review by gizmocuz — COMMENTED** [review]
A few comments

**Comment by gizmocuz** [comment]
Is it possible to make the above changes ?

**Comment by LemarinelNet** [comment]
Yes , I'll commit them tomorrow.

**Comment by soif** [comment]
It would be amazing @LemarinelNet, This would allow to use dimmer messages, to send a number between 0-100 (ie send a MODE/PRESET/NUMBER to a Mysensor node). 

I cant wait for this excellent PR ;-) !

**Comment by LemarinelNet** [comment]
Glad to see that this PR will be usefull ;)
I've made the modifications, still testing them ;)

**Comment by soif** [comment]
de la balle ;-)

**Comment by LemarinelNet** [comment]
Made the needed changes.
Performed a bunch of tests.
I've commited the changes !

File: hardware/MySensorsBase.cpp [hunk 4]
```diff
@@ -1599,7 +1602,7 @@ void MySensorsBase::ParseLine()
 		return;
 	std::string sLine((char*)&m_buffer);
 
-	//_log.Log(LOG_STATUS, sLine.c_str());
+	_log.Log(LOG_STATUS, "MySensors: frame received : %s", sLine.c_str());
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp]: Could you remark this log line ?

File: hardware/MySensorsBase.cpp [hunk 5]
```diff
@@ -1146,11 +1149,15 @@ void MySensorsBase::UpdateSwitch(const _eSetType vType, const unsigned char Idx,
...
+	sprintf(szIdx, "%02X%02X%02X%02X", 0, 0, 0, Idx);
 	std::vector<std::vector<std::string> > result;
-	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeGeneralSwitch), int(sTypeAC));
+
+	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%s') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeGeneralSwitch), int(sTypeAC));
+
+
 	if (!result.empty())
 	{
+		_log.Log(LOG_ERROR, "Result not empty. " );
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp]: Could you remark this log line ?

**Comment by LemarinelNet** [comment]
Oups ! 
I was using it to debug, I remove it now ;)

**Comment by gizmocuz** [comment]
Thanks!

**Comment by gizmocuz** [comment]
Is it possible to have a look on our forum, there seems to be a problem, and it could be related to this patch:
https://www.domoticz.com/forum/viewtopic.php?f=6&t=16799

**Comment by LemarinelNet** [comment]
I think that it is because the node has been created with a previous version, so created as Lighting2 node, and then when it is triggered from new update, the node sends a Lighting2 type, which is not handled anymore.
Perhaps I should add retro-compatible code in the WriteToHardware function ?

**Comment by gizmocuz** [comment]
Best would have been to make a database version increase and a patch in SQLHelper.cpp so all old sensors where converted to the new type
Maybe you can do this ? It will not help current beta users, but for people on the stable version it would help

File: hardware/MySensorsBase.cpp [hunk 6]
```diff
@@ -1169,29 +1176,27 @@ void MySensorsBase::UpdateSwitch(const _eSetType vType, const unsigned char Idx,
...
-	lcmd.LIGHTING2.subtype = sTypeAC;
-	lcmd.LIGHTING2.id1 = 0;
-	lcmd.LIGHTING2.id2 = 0;
-	lcmd.LIGHTING2.id3 = 0;
-	lcmd.LIGHTING2.id4 = Idx;
-	lcmd.LIGHTING2.unitcode = SubUnit;
+	// LLEMARINEL : #1312  Changed to use as pTypeGeneralSwitch
+	// Send as General Switch :
+	_tGeneralSwitch gswitch;
+	gswitch.subtype = sTypeAC;
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp:1182]: Help! This is the wrong sub type, should not be sTypeAC, have a look in hardwaretypes.h where pTypeGeneralSwitch is defined!
I will create a patch now

File: hardware/MySensorsBase.cpp [hunk 7]
```diff
@@ -1141,13 +1145,16 @@ void MySensorsBase::UpdateRGBWSwitchLastUpdate(const int NodeID, const int Child
...
+	// LLEMARINEL : #1312  Changed to use as pTypeGeneralSwitch : do not constrain to 16 steps anymore but 100 :
+	int level = int(Level);
 
 	char szIdx[10];
-	sprintf(szIdx, "%X%02X%02X%02X", 0, 0, 0, Idx);
+	sprintf(szIdx, "%02X%02X%02X%02X", 0, 0, 0, Idx);
 	std::vector<std::vector<std::string> > result;
-	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%q') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeLighting2), int(sTypeAC));
+
+	result = m_sql.safe_query("SELECT Name,nValue,sValue FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%s') AND (Unit==%d) AND (Type==%d) AND (Subtype==%d)", m_HwdID, szIdx, SubUnit, int(pTypeGeneralSwitch), int(sTypeAC));
```
Comment (gizmocuz) [inline hardware/MySensorsBase.cpp:1155]: Help! This is the wrong sub type, should not be sTypeAC, have a look in hardwaretypes.h where pTypeGeneralSwitch is defined!
I will create a patch now

**Review by gizmocuz — COMMENTED** [review]
Wrong Subtype for switch used, creating patch now

**Comment by gizmocuz** [comment]
Patch made
https://github.com/domoticz/domoticz/commit/345f92cd0ab0f0f0c5762646d704905ad088348a

**Pull request merged**